libraries/CurieI2S/src/CurieI2S.cpp

```diff
@@ -1,125 +1,270 @@
+//***************************************************************
+//
+// Copyright (c) 2016 Intel Corporation.  All rights reserved.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+//
+//***************************************************************
+
 //CurieI2S.cpp
 
 #include "CurieI2S.h"
 #include <interrupt.h>
 
 Curie_I2S CurieI2S;
 
-static uint32_t _i2s_dma_Rx_Buffer[DMA_BUFFER_SIZE];
-static uint32_t _i2s_dma_Tx_Buffer[DMA_BUFFER_SIZE];
+//static uint32_t _i2s_dma_Rx_Buffer[DMA_BUFFER_SIZE];
+//static uint32_t _i2s_dma_Tx_Buffer[DMA_BUFFER_SIZE];
 
 static struct i2s_ring_buffer _i2s_Rx_Buffer;
 static struct i2s_ring_buffer _i2s_Tx_Buffer;
 
 static struct i2s_ring_buffer *_i2s_Rx_BufferPtr = &_i2s_Rx_Buffer;
 static struct i2s_ring_buffer *_i2s_Tx_BufferPtr = &_i2s_Tx_Buffer;
 
+//static int _i2s_frame_delay = 960;
+
 static void i2sInterruptHandler(void)
 {
-    //TODO: Make interrupt handling more atomic
-    //noInterrupts();
-    //uint32_t fifoL = *I2S_TFIFO_STAT;
+    //Serial.println("i2s int");
+    uint32_t i2s_stat = *I2S_STAT;
+    
+    //RX FIFO almost full
+    if((*I2S_STAT & 0x00008000) && (*I2S_CID_CTRL & 0x80000000))
+    {
+        //Serial.println("rx almost full");
+        
+        //disable RFIFO_AFULL interrupts
+        *I2S_CID_CTRL = *I2S_CID_CTRL & 0x7FFFFFFF;
+        
+        int index;
+        int fifoDataLength = (*I2S_RFIFO_STAT & 0x0000000F);
+
+        for(int i = 0; i < fifoDataLength; i++)
+        {
+            index = (uint32_t)(_i2s_Rx_BufferPtr->head +1) % I2S_BUFFER_SIZE;
+            uint32_t data = *I2S_DATA_REG;
+            if(index != _i2s_Rx_BufferPtr->tail)
+            {
+                _i2s_Rx_BufferPtr->data[_i2s_Rx_BufferPtr->head] = data;
+                _i2s_Rx_BufferPtr->head = index;
+            }
+            #ifdef I2S_DEBUG
+            digitalWrite(I2S_DEBUG_PIN, HIGH);
+            digitalWrite(I2S_DEBUG_PIN, LOW);
+            #endif
+
+        }
+        
```

Comment:
May need to consider a timing window here.  Consider a byte came in, after fifoDataLength was set with the h/w register earlier on.  The FIFO would not be emptied here.  With the "Almost Full" interrupt disabled and the "Empty" interrupt enabled, there may no more interrupt from here on.  Would that be a real case?